### PR TITLE
Improve certificate performance

### PR DESCRIPTION
**Pull request by lehins**
# Description

This PR improves certificate performance and makes it more clear what is going on.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs [hunk 1]
```diff
@@ -193,22 +192,22 @@ conwayDelegTransition = do
       checkStakeKeyNotRegistered stakeCred
       pure $ dState {dsUnified = registerStakeCredential stakeCred}
     ConwayUnRegCert stakeCred sMayRefund -> do
-      let mRDPair = UM.lookup stakeCred $ UM.RewDepUView dsUnified
+      let (mUMElem, umap) = UM.extractStakingCredential stakeCred dsUnified
```
Comment (TimSheard) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:195]: Add a comment some thing like this

The UMap has two internal Maps.  umElems and umPtrs. The operation extractStakingCredential
completely removes the credential from the UmElems map, and also returns the  UMElem associated with that Credential. By examining that UMElem we can construct tests that ensure valid Delegation. It also adjusts the umPtrs map, efficiently removing pairs from that map that are associated with that Credential.

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs [hunk 2]
```diff
@@ -276,19 +283,16 @@ delegationTransition = do
...
-      let rewardCoin = rdReward <$> UM.lookup hk (rewards ds)
-      rewardCoin == Just mempty ?! StakeKeyNonZeroAccountBalanceDELEG (fromCompact <$> rewardCoin)
-
-      let u0 = dsUnified ds
-          u1 = Set.singleton hk UM.⋪ RewDepUView u0
-          u2 = Set.singleton hk UM.⋪ SPoolUView u1
-          u3 = PtrUView u2 UM.⋫ Set.singleton hk
-          u4 = ds {dsUnified = u3}
-      pure u4
+      let (mUMElem, umap) = UM.extractStakingCredential cred (dsUnified ds)
```
Comment (TimSheard) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:288]: Repeat the comment here

The UMap has two internal Maps. umElems and umPtrs. The operation extractStakingCredential
completely removes the credential from the UmElems map, and also returns the UMElem associated with that Credential. By examining that UMElem we can construct tests that ensure valid Delegation. It also adjusts the umPtrs map, efficiently removing pairs from that map that are associated with that Credential..

File: libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs [hunk 3]
```diff
@@ -944,11 +946,25 @@ domDelete = (⋪)
...
-    , umPtrs = Map.filter (`Set.notMember` ks) umPtrs
-    }
+domDeleteAll ks umap = Set.foldr' deleteStakingCredential umap ks
+
+-- | Completely remove the staking credential from the UMap, including all associated
+-- pointers.
+deleteStakingCredential :: Credential 'Staking c -> UMap c -> UMap c
+deleteStakingCredential cred = snd . extractStakingCredential cred
+
+-- | Just like `deleteStakingCredential`, but also returned the removed element.
```
Comment (TimSheard) [inline libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:956]: The operation extractStakingCredential completely removes the credential from the UmElems map, and also returns the UMElem associated with that Credential (useful for performing predicate failure tests)  It also adjusts the umPtrs map, efficiently removing pairs from that map that are associated with that Credential.

**Review by TimSheard — APPROVED** [review]
Just a few comments to add to the code.

**Review by Lucsanszky — APPROVED** [review]

**Review by teodanciu — APPROVED** [review]
Looks good to me, we could later add `  deleteStakingCredential` and  `extractStakingCredential` to cardano-ledger-core CHANGELOG.

**Pull request merged**